spec runner depends on libxml

To run specs, you need to require `spec` which includes `junit_formatter`. That formatter requires `xml` which depends on `libxml2` being available. That package is not installed by default - at least not on Debian/Ubuntu or homebrew (the APT package has it as a a suggested dependecy). Didn't check others but I don't expect them to be different.

This means, right after installing Crystal using the provided packages, you're not able to run specs. Instead the compiler will print a very cryptic error message (like posted in #5658):

```
/usr/bin/ld: cannot find -lxml2
collect2: error: ld returned 1 exit status
Error: execution of command failed with code: 1: `cc "${@}" -o '/home/Mead/.cache/crystal/crystal-run-spec.tmp'  -rdynamic  -lxml2 -lpcre -lm -lgc -lpthread /usr/share/crystal/src/ext/libcrystal.a -levent -lrt -ldl -L/usr/lib -L/usr/local/lib`
```

For beginners, this is not trivial to understand. The error message could obviously be improved (see #5291) but that won't solve the problem that you can't use a major feature without manually installing additional libraries.

This is especially iffy considering the JUnit formatter is a very specific feature and won't be needed for most use cases at all.

A possible solution would be to include `libxml2` in the dependencies of the OS packages for Crystal. But I don't think that's the best way because many developers probably won't really use xml anyway so it'd just add an unneccesary dependency.

I think it would be better to make the junit formatter optional. Specs are usually compiled and run using `crystal spec`, so it should be possible to only require `junit_formatter` in the code being compiled if it is configured to use that formatter (as CLI argument).

Or remove it entirely from the stdlib into a shard as suggested in #5215.

Related to #5291

--- a/spec/std/spec/junit_formatter_spec.cr
+++ b/spec/std/spec/junit_formatter_spec.cr
@@ -9,15 +9,14 @@ describe "JUnit Formatter" do
     end
 
     expected = <<-XML
-                 <testsuite tests="2" errors="0" failed="0">
-                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something">
-                 </testcase>
-                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something else">
-                 </testcase>
+                 <?xml version="1.0"?>
+                 <testsuite tests="2" errors="0" failures="0">
+                   <testcase file="spec/some_spec.cr" classname="spec.some_spec" name="should do something"/>
+                   <testcase file="spec/some_spec.cr" classname="spec.some_spec" name="should do something else"/>
                  </testsuite>
                  XML
 
-    output.should eq(expected)
+    output.chomp.should eq(expected)
   end
 
   it "reports failures" do
@@ -26,14 +25,15 @@ describe "JUnit Formatter" do
     end
 
     expected = <<-XML
-                 <testsuite tests="1" errors="0" failed="1">
-                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something">
-                 <failure />
-                 </testcase>
+                 <?xml version="1.0"?>
+                 <testsuite tests="1" errors="0" failures="1">
+                   <testcase file="spec/some_spec.cr" classname="spec.some_spec" name="should do something">
+                     <failure/>
+                   </testcase>
                  </testsuite>
                  XML
 
-    output.should eq(expected)
+    output.chomp.should eq(expected)
   end
 
   it "reports errors" do
@@ -42,14 +42,15 @@ describe "JUnit Formatter" do
     end
 
     expected = <<-XML
-                 <testsuite tests="1" errors="1" failed="0">
-                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something">
-                 <error />
-                 </testcase>
+                 <?xml version="1.0"?>
+                 <testsuite tests="1" errors="1" failures="0">
+                   <testcase file="spec/some_spec.cr" classname="spec.some_spec" name="should do something">
+                     <error/>
+                   </testcase>
                  </testsuite>
                  XML
 
-    output.should eq(expected)
+    output.chomp.should eq(expected)
   end
 
   it "reports mixed results" do
@@ -61,31 +62,31 @@ describe "JUnit Formatter" do
     end
 
     expected = <<-XML
-                 <testsuite tests="4" errors="2" failed="1">
-                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something1">
-                 </testcase>
-                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something2">
-                 <failure />
-                 </testcase>
-                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something3">
-                 <error />
-                 </testcase>
-                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something4">
-                 <error />
-                 </testcase>
+                 <?xml version="1.0"?>
+                 <testsuite tests="4" errors="2" failures="1">
+                   <testcase file="spec/some_spec.cr" classname="spec.some_spec" name="should do something1"/>
+                   <testcase file="spec/some_spec.cr" classname="spec.some_spec" name="should do something2">
+                     <failure/>
+                   </testcase>
+                   <testcase file="spec/some_spec.cr" classname="spec.some_spec" name="should do something3">
+                     <error/>
+                   </testcase>
+                   <testcase file="spec/some_spec.cr" classname="spec.some_spec" name="should do something4">
+                     <error/>
+                   </testcase>
                  </testsuite>
                  XML
 
-    output.should eq(expected)
+    output.chomp.should eq(expected)
   end
 
   it "escapes spec names" do
     output = build_report do |f|
-      f.report Spec::Result.new(:success, "complicated \" <n>'&ame", __FILE__, __LINE__, nil, nil)
+      f.report Spec::Result.new(:success, %(complicated " <n>'&ame), __FILE__, __LINE__, nil, nil)
     end
 
     name = XML.parse(output).xpath_string("string(//testsuite/testcase[1]/@name)")
-    name.should eq("complicated \" <n>'&ame")
+    name.should eq(%(complicated \" <n>'&ame))
   end
 
   it "report failure stacktrace if present" do

--- a/spec/std/spec/junit_formatter_spec.cr
+++ b/spec/std/spec/junit_formatter_spec.cr
@@ -1,4 +1,5 @@
 require "spec"
+require "xml"
 
 describe "JUnit Formatter" do
   it "reports successful results" do
@@ -8,14 +9,15 @@ describe "JUnit Formatter" do
     end
 
     expected = <<-XML
-                 <?xml version="1.0"?>
-                 <testsuite tests="2" errors="0" failures="0">
-                   <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something"/>
-                   <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something else"/>
+                 <testsuite tests="2" errors="0" failed="0">
+                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something">
+                 </testcase>
+                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something else">
+                 </testcase>
                  </testsuite>
                  XML
 
-    output.chomp.should eq(expected)
+    output.should eq(expected)
   end
 
   it "reports failures" do
@@ -24,15 +26,14 @@ describe "JUnit Formatter" do
     end
 
     expected = <<-XML
-                 <?xml version="1.0"?>
-                 <testsuite tests="1" errors="0" failures="1">
-                   <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something">
-                     <failure/>
-                   </testcase>
+                 <testsuite tests="1" errors="0" failed="1">
+                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something">
+                 <failure />
+                 </testcase>
                  </testsuite>
                  XML
 
-    output.chomp.should eq(expected)
+    output.should eq(expected)
   end
 
   it "reports errors" do
@@ -41,15 +42,14 @@ describe "JUnit Formatter" do
     end
 
     expected = <<-XML
-                 <?xml version="1.0"?>
-                 <testsuite tests="1" errors="1" failures="0">
-                   <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something">
-                     <error/>
-                   </testcase>
+                 <testsuite tests="1" errors="1" failed="0">
+                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something">
+                 <error />
+                 </testcase>
                  </testsuite>
                  XML
 
-    output.chomp.should eq(expected)
+    output.should eq(expected)
   end
 
   it "reports mixed results" do
@@ -61,22 +61,22 @@ describe "JUnit Formatter" do
     end
 
     expected = <<-XML
-                 <?xml version="1.0"?>
-                 <testsuite tests="4" errors="2" failures="1">
-                   <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something1"/>
-                   <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something2">
-                     <failure/>
-                   </testcase>
-                   <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something3">
-                     <error/>
-                   </testcase>
-                   <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something4">
-                     <error/>
-                   </testcase>
+                 <testsuite tests="4" errors="2" failed="1">
+                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something1">
+                 </testcase>
+                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something2">
+                 <failure />
+                 </testcase>
+                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something3">
+                 <error />
+                 </testcase>
+                 <testcase file=\"spec/some_spec.cr\" classname=\"spec.some_spec\" name="should do something4">
+                 <error />
+                 </testcase>
                  </testsuite>
                  XML
 
-    output.chomp.should eq(expected)
+    output.should eq(expected)
   end
 
   it "escapes spec names" do

--- a/src/spec/junit_formatter.cr
+++ b/src/spec/junit_formatter.cr
@@ -71,8 +71,9 @@ module Spec
       if message = exception.message
         io << %( message=")
         HTML.escape(message, io)
+        io << '"'
       end
-      io << %(">)
+      io << '>'
 
       if backtrace = exception.backtrace?
         HTML.escape(backtrace.join('\n'), io)

--- a/src/spec/junit_formatter.cr
+++ b/src/spec/junit_formatter.cr
@@ -10,15 +10,6 @@ module Spec
     def initialize(@output)
     end
 
-    def push(context)
-    end
-
-    def pop
-    end
-
-    def before_example(description)
-    end
-
     def report(result)
       current = @summary[result.kind]? || 0
       @summary[result.kind] = current + 1
@@ -27,13 +18,16 @@ module Spec
 
     def finish
       io = @output
-      io << "<testsuite tests=\"#{@results.size}\" \
-                        errors=\"#{@summary[:error]? || 0}\" \
-                        failed=\"#{@summary[:fail]? || 0}\">\n"
+      io.puts %(<?xml version="1.0"?>)
+      io << %(<testsuite tests=") << @results.size
+      io << %(" errors=") << (@summary[:error]? || 0)
+      io << %(" failures=") << (@summary[:fail]? || 0) << %(">)
+
+      io.puts
 
       @results.each { |r| write_report(r, io) }
 
-      io << "</testsuite>"
+      io << %(</testsuite>)
       io.close
     end
 
@@ -46,44 +40,45 @@ module Spec
 
     # -------- private utility methods
     private def write_report(result, io)
-      io << "<testcase file=\"#{result.file}\" classname=\"#{classname(result)}\" name=\"#{HTML.escape(result.description)}\">\n"
+      io << %(  <testcase file=") << result.file
+      io << %(" classname=") << classname(result) << %(" name=")
+      HTML.escape(result.description, io)
 
-      if (has_inner_content(result.kind))
-        tag = inner_content_tag(result.kind)
-        ex = result.exception
-        if ex
-          write_inner_content(tag, ex, io)
+      if tag = inner_content_tag(result.kind)
+        io.puts %(">)
+
+        if exception = result.exception
+          write_inner_content(tag, exception, io)
         else
-          io << "<#{tag} />\n"
+          io << "    <" << tag << "/>\n"
         end
+        io.puts "  </testcase>"
+      else
+        io.puts %("/>)
       end
-
-      io << "</testcase>\n"
-    end
-
-    private def has_inner_content(kind)
-      kind == :fail || kind == :error
     end
 
     private def inner_content_tag(kind)
       case kind
-      when :error
-        :error
-      when :fail
-        :failure
+      when :error then "error"
+      when :fail  then "failure"
       end
     end
 
     private def write_inner_content(tag, exception, io)
-      m = exception.message
-      if m
-        io << "<#{tag} message=\"#{HTML.escape(m)}\">"
-      else
-        io << "<#{tag}>"
+      io << "    <" << tag
+
+      if message = exception.message
+        io << %( message=")
+        HTML.escape(message, io)
       end
-      backtrace = exception.backtrace? || ([] of String)
-      io << HTML.escape(backtrace.join("\n"))
-      io << "</#{tag}>\n"
+      io << %(">)
+
+      if backtrace = exception.backtrace?
+        HTML.escape(backtrace.join('\n'), io)
+      end
+
+      io << "</" << tag << ">\n"
     end
 
     private def classname(result)

--- a/src/spec/junit_formatter.cr
+++ b/src/spec/junit_formatter.cr
@@ -1,4 +1,4 @@
-require "xml"
+require "html"
 
 module Spec
   # :nodoc:
@@ -27,20 +27,14 @@ module Spec
 
     def finish
       io = @output
+      io << "<testsuite tests=\"#{@results.size}\" \
+                        errors=\"#{@summary[:error]? || 0}\" \
+                        failed=\"#{@summary[:fail]? || 0}\">\n"
 
-      XML.build(io, indent: 2) do |xml|
-        attributes = {
-          tests:    @results.size,
-          errors:   @summary[:error]? || 0,
-          failures: @summary[:fail]? || 0,
-        }
+      @results.each { |r| write_report(r, io) }
 
-        xml.element("testsuite", attributes) do
-          @results.each { |r| write_report(r, xml) }
-        end
-      end
-    ensure
-      io.try &.close
+      io << "</testsuite>"
+      io.close
     end
 
     def self.file(output_dir)
@@ -50,44 +44,46 @@ module Spec
       JUnitFormatter.new(file)
     end
 
-    private def write_report(result, xml)
-      attributes = {
-        file:      result.file,
-        classname: classname(result),
-        name:      result.description,
-      }
+    # -------- private utility methods
+    private def write_report(result, io)
+      io << "<testcase file=\"#{result.file}\" classname=\"#{classname(result)}\" name=\"#{HTML.escape(result.description)}\">\n"
 
-      xml.element("testcase", attributes) do
-        if tag = inner_content_tag(result.kind)
-          if ex = result.exception
-            write_inner_content(tag, ex, xml)
-          else
-            xml.element(tag)
-          end
+      if (has_inner_content(result.kind))
+        tag = inner_content_tag(result.kind)
+        ex = result.exception
+        if ex
+          write_inner_content(tag, ex, io)
+        else
+          io << "<#{tag} />\n"
         end
       end
+
+      io << "</testcase>\n"
+    end
+
+    private def has_inner_content(kind)
+      kind == :fail || kind == :error
     end
 
     private def inner_content_tag(kind)
       case kind
       when :error
-        "error"
+        :error
       when :fail
-        "failure"
+        :failure
       end
     end
 
-    private def write_inner_content(tag, exception, xml)
-      if message = exception.message
-        attributes = {message: message}
+    private def write_inner_content(tag, exception, io)
+      m = exception.message
+      if m
+        io << "<#{tag} message=\"#{HTML.escape(m)}\">"
       else
-        attributes = NamedTuple.new
+        io << "<#{tag}>"
       end
-
-      xml.element(tag, attributes) do
-        backtrace = exception.backtrace? || Array(String).new
-        xml.text backtrace.join('\n')
-      end
+      backtrace = exception.backtrace? || ([] of String)
+      io << HTML.escape(backtrace.join("\n"))
+      io << "</#{tag}>\n"
     end
 
     private def classname(result)

--- a/src/spec/junit_formatter.cr
+++ b/src/spec/junit_formatter.cr
@@ -40,8 +40,11 @@ module Spec
 
     # -------- private utility methods
     private def write_report(result, io)
-      io << %(  <testcase file=") << result.file
-      io << %(" classname=") << classname(result) << %(" name=")
+      io << %(  <testcase file=")
+      HTML.escape(result.file, io)
+      io << %(" classname=")
+      HTML.escape(classname(result), io)
+      io << %(" name=")
       HTML.escape(result.description, io)
 
       if tag = inner_content_tag(result.kind)

